tests: Check PIM Register/-Stop handling in pim_igmp_vrf topotest

gromit1811 opened the pull request:
Check whether PIM Register messages are generated towards the RP and answered by Register-Stop from the RP. To force generation of PIM Register, move the DR role from the RP on R11/12 to R1 (otherwise, with DR and RP being the same routers, no Register messages would be needed). The RPs need to generate Register-Stop because there are no other group members besides the ones on the path whether the traffic is received from on the RPs.

Useful as a regression test for https://github.com/FRRouting/frr/pull/18216 even though the situation there was slightly different: In that case, the RP was the VRF router and its Register-Stop messages ended up the in the wrong VRF. In our case, the DRs are the VRF routers and the Register messages would end up in the wrong VRF. But our check works for both cases: If we don't receive Register-Stop messages, either the Register or the Register-Stop messages got lost and will trigger an assertion failure.

Note: The last commit is the actual change to check PIM Register. The other commits are just minor cleanup/bugfix changes to the original topotest.

gromit1811 commented:
Converted to draft while working on IPv6 support

Jafaral commented on tests/topotests/pim_igmp_vrf/test_pim_vrf.py:
```diff
@@ -323,7 +323,7 @@ def _test_vrf_pimreg_interfaces():
     r1 = tgen.gears["r1"]
     r1.vtysh_cmd("conf\ninterface blue\nip pim")
-    r1.vtysh_cmd("conf\nvrf blue\nip pim rp 192.168.0.11 239.100.0.1/32\nexit-vrf")
+    r1.vtysh_cmd("conf\nrouter pim vrf blue\nrp 192.168.0.11 239.100.0.1/32")
```
While you are at it, can you please use the following more readable format
```
r1.vtysh_cmd(
"""
conf
router pim vrf blue
   rp 192.168.0.11 239.100.0.1/32
"""
)

```

gromit1811 replied:
Will also merge the 2 vtysh invocations. When using multi-line, there's no point in calling vtysh_cmd() twice.

Jafaral commented on tests/topotests/pim_igmp_vrf/test_pim_vrf.py:
```diff
@@ -339,7 +339,7 @@ def _test_vrf_pimreg_interfaces():
 
     r1.vtysh_cmd("conf\ninterface red\nip pim")
-    r1.vtysh_cmd("conf\nvrf red\nip pim rp 192.168.0.12 239.100.0.1/32\nexit-vrf")
+    r1.vtysh_cmd("conf\nrouter pim vrf red\nrp 192.168.0.12 239.100.0.1/32")
```
same comment above.

gromit1811 replied:
Actually no longer exists. I'm now using one format string with placeholders for "red" and "blue" and the IPs.

Jafaral commented on tests/topotests/pim_igmp_vrf/r1/pimd.conf:
```diff
@@ -9,6 +9,12 @@ hostname r1
 ! debug pim zebra
 ! debug pim bsm
 !
+vrf blue
+exit-vrf
+!
+vrf red
+exit-vrf
+!
```
While I agree that we should fix this. A more permanent fix for this test is to actually reflect the interface vrf assignment below. I never liked relying on side effects from the kernel to map interfaces to vrfs because the configs don't really tell us much about that. Here is the the vrf info from the test file. 
```
        "ip link set dev r1-eth0 vrf blue up",
        "ip link set dev r1-eth1 vrf blue up",
        "ip link set dev r1-eth2 vrf red up",
        "ip link set dev r1-eth3 vrf red up",
```
So for interfaces below you would have
```
interface r1-eth0 vrf blue
```
That is the correct config that will behave exactly as intended and also reflects the mapping for interfaces to vrfs.

Jafaral replied:
That is probably a race condition. The kernel sends us an  new interface message say new `eth0`. If there is no vrf binding in the kernel for that interface, then it will show without a vrf in the configs, or if the message arrived late.

gromit1811 replied:
Just checked this again. It even happens when I interactively configure this using vtysh, so it's unlikely to be a race condition (I don't type _that_ fast :wink:)

For instance with FRR from git master 361f80a64b69640dbbf472e77d6d04d2e62f409a plus this PR, i.e. post-vrf-fix. System VRF config:
```root@debian:~/frr# ip -d l sh dev ovl2336
504: ovl2336: <NOARP,MASTER,UP,LOWER_UP> mtu 65575 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 52:49:4a:dd:27:aa brd ff:ff:ff:ff:ff:ff promiscuity 0 minmtu 1280 maxmtu 65575 
    vrf table 2336 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
root@debian:~/frr# ip -d l sh dev veth0a
512: veth0a@veth0b: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master ovl2336 state UP mode DEFAULT group default qlen 1000
    link/ether 12:36:87:5a:e2:f2 brd ff:ff:ff:ff:ff:ff promiscuity 0 minmtu 68 maxmtu 65535 
    veth 
    vrf_slave table 2336 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
```
-> VRF `ovl2336` plus interface `veth0a` in this VRF.

Trying to configure these interactively in FRR with initial config empty:
```
debian# show run
Building configuration...

Current configuration:
!
frr version 10.4-dev-my-manual-build
frr defaults traditional
hostname debian
no ip forwarding
no ipv6 forwarding
service integrated-vtysh-config
!
end
debian# conf
debian(config)# vrf ovl2336 
debian(config-vrf)# exit-vrf 
debian(config)# interface veth0a vrf ovl2336 
debian(config-if)# ip pim
debian(config-if)# exit
debian(config)# router pim vrf ovl2336 
debian(config-pim)# join-prune-interval 5
debian(config-pim)# end
debian# show run
Building configuration...

Current configuration:
!
frr version 10.4-dev-my-manual-build
frr defaults traditional
hostname debian
no ip forwarding
no ipv6 forwarding
service integrated-vtysh-config
!
vrf ovl2336
exit-vrf
!
interface veth0a
 ip pim
exit
!
router pim
 join-prune-interval 5
exit
!
end
```

-> All my config commands are there, but all `vrf ovl2336` arguments have been stripped. Am I doing something wrong? Am I expecting too much?

gromit1811 replied:
Should I open yet another issue for that?

Jafaral reviewed (changes requested):
Thank you @gromit1811

Jafaral commented on tests/topotests/pim_igmp_vrf/r1/pimd.conf:
```diff
@@ -9,20 +9,33 @@ hostname r1
 ! debug pim bsm
 !
-interface r1-eth0
+vrf blue
+exit-vrf
```
Can you please open an issue to track this one? will try to find time to fix it.

gromit1811 replied:
Can't reproduce this right now with the current version of the testcase with unified config. Will try to find an easy way how to reproduce this and then open an issue.

gromit1811 replied:
OK, also happens with unified config as long as you don't load it using `router.load_frr_config()` but like before with per-daemon calls to `router.load_config()`

gromit1811 replied:
[#18429](https://github.com/FRRouting/frr/issues/18429)

Jafaral commented:
new format suggestion as well.

gromit1811 commented on tests/topotests/pim_igmp_vrf/r1/pim6_blue_join.json:
```diff
@@ -0,0 +1,22 @@
+  "r1-eth0":{
+    "name":"r1-eth0",
+    "state":"up",
+    "address":"fe80::XXXX:XXXX:XXXX:XXXX",
+    "flagMulticast":true,
+    "flagBroadcast":true,
+    "lanDelayEnabled":true,
+    "ff18:100::1":{
+      "*":{
+        "source":"*",
+        "group":"ff18:100::1",
+        "upTime":"--:--:--",
+        "expire":"--:--",
+        "prune":"--:--",
```
Without fully understanding what this test is trying to test, I didn't want to deviate too much from the original and just did things the same way as in IPv4.

Jafaral commented:
@gromit1811 if you rebase , you no longer need to do the funny vrf business. Fixed on master.

gromit1811 commented:
Rebased to get the vrf fix.

Added yet another test case to check interface/router removal. PIM interface removal seems to work properly, but completely disabling PIM (e.g. causing the pimreg interface to disappear) doesn't. Also, "no router pim vrf XXX" doesn't actually remove the PIM section from config, "show run" afterwards still shows it.

I guess that's not expected behavior, right? I'll open an issue for that unless somebody tells me it's OK like this.

Jafaral commented:
> Rebased to get the vrf fix.
> 
> Added yet another test case to check interface/router removal. PIM interface removal seems to work properly, but completely disabling PIM (e.g. causing the pimreg interface to disappear) doesn't. Also, "no router pim vrf XXX" doesn't actually remove the PIM section from config, "show run" afterwards still shows it.
> 
> I guess that's not expected behavior, right? I'll open an issue for that unless somebody tells me it's OK like this.

Yeah, that should probably be fixed.

gromit1811 commented:
> Yeah, that should probably be fixed.

Not fixed, but at least reported: https://github.com/FRRouting/frr/issues/18537

gromit1811 commented:
Pushed an update with some pytest mark fixes and a modified join/MC stream start order: Join first, then start MC stream 1 second later (before: start MC stream and then immediately join). With this, #18445 now also occurs for IPv4 :frowning_face:

gromit1811 commented:
No autoclose please, #18445 is still open, so this should remain as well

gromit1811 commented:
Rebased, updated to cope with new OSPFv3 JSON route format, dropped empty VRF config to work around assertion failure (no longer needed with current master), removed new cleanup test cases (will open a separate PR so we can finally merge this one).
This now passes with latest master thanks to f41e730.
